packages/amplify-e2e-tests/src/__tests__/version-gating.test.ts

```diff
@@ -0,0 +1,232 @@
+
+  const creds = new aws.SharedIniFileCredentials({ profile: profileName });
+  aws.config.credentials = creds;
+
+  let region: string;
+  let rootStackName: string;
+  let cfnClient: CloudFormation;
+
+  beforeEach(async () => {
+    projRoot = await createNewProjectDir(projName);
+
+    if (!isPackagedAmplifyInPath()) {
+      cliPath = resolveRealCLIPath();
+      packageJsonPath = path.resolve(cliPath, '..', '..', '..', 'amplify-cli', 'package.json');
+      originalPackageJsonContent = fs.readFileSync(packageJsonPath, 'utf8').toString();
+      packageJson = JSONUtilities.parse<{ version: string; amplifyCLIConfiguration: { minimumCLIVersion: string } }>(
+        originalPackageJsonContent,
+      );
+    }
+  });
+
+  afterEach(async () => {
+    await deleteProject(projRoot);
+    deleteProjectDir(projRoot);
+  });
+
+  it('a', async () => {
+    // We cannot execute this test for packaged CLI as we are modifying the package.json file to
+    // simulate multiple versions
+    if (isPackagedAmplifyInPath()) {
+      return;
+    }
+
+    try {
+      // Reset version information to make sure tests will pass with future CLI versionsas well
+      updateVersionInPackageJson(baseCLIVersion, baseMinimumCLIVersion);
+
+      await initJSProjectWithProfile(projRoot, { name: projName });
+
+      const meta = stateManager.getMeta(projRoot);
+
+      region = _.get(meta, ['providers', 'awscloudformation', 'Region']);
+      rootStackName = _.get(meta, ['providers', 'awscloudformation', 'StackName']);
+
+      cfnClient = new CloudFormation({ region });
+
+      await expectStackMetadata(baseCLIVersion, baseMinimumCLIVersion);
+
+      // Set CLI version to 100.1.0, verify push succeeds with newer version of CLI
+      updateVersionInPackageJson('100.1.0', undefined);
+
+      await push();
+
+      await expectStackMetadata('100.1.0', baseMinimumCLIVersion);
+
+      // Set Minimum CLI version to 100.2.0, verify push succeeds with newer version of CLI
+      updateVersionInPackageJson(undefined, '100.2.0');
+
+      await pushFail();
+
+      // Set CLI version to 100.2.0, Minimum CLI version to 100.2.0, verify push succeeds
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+
+      // Set CLI version to 100.3.0, Minimum CLI version to 100.2.0, verify push succeeds
+      updateVersionInPackageJson('100.3.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.3.0', '100.2.0');
+
+      // Set CLI version back to 100.2.0, Minimum CLI version to 100.2.0, verify push succeeds as
+      // minimum deploy version requirement has met
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+    } finally {
+      // restore original package.json
+      fs.writeFileSync(packageJsonPath, originalPackageJsonContent);
```

Comment:
Maybe this could go in the `afterEach()` in case more tests are added in the future.

Reply:
the reason why it is in finally is that we want to restore it no matter what in case of error as well and scoping it here makes more sense imho